pkg/storage/scrub.go

```diff
@@ -92,84 +88,171 @@ func CheckRepo(ctx context.Context, imageName string, imgStore storageTypes.Imag
-		if manifest.MediaType == ispec.MediaTypeImageIndex {
-			buf, err := os.ReadFile(path.Join(dir, "blobs", manifest.Digest.Algorithm().String(), manifest.Digest.Encoded()))
-			if err != nil {
-				tagName := manifest.Annotations[ispec.AnnotationRefName]
-				imgRes := getResult(imageName, tagName, errors.ErrBadBlobDigest)
-				results = append(results, imgRes)
-
-				continue
-			}
+		tag := manifest.Annotations[ispec.AnnotationRefName]
+		imageResult := getManifestResult(ctx, manifest, imgStore, imageName, tag, listOfManifests)
+		results = append(results, imageResult)
+	}
+
+	return results, nil
+}
+
+func getManifestResult(
+	ctx context.Context, manifest ispec.Descriptor, imgStore storageTypes.ImageStore, imageName, tag string,
+	listOfManifests map[godigest.Digest]ScrubImageResult,
+) ScrubImageResult {
+	res, ok := listOfManifests[manifest.Digest]
+	if ok {
+		return ScrubImageResult{
+			ImageName: imageName,
+			Tag:       tag,
+			Status:    res.Status,
+			Error:     res.Error,
+		}
+	}
+
+	switch manifest.MediaType {
+	case ispec.MediaTypeImageIndex:
+		buf, err := imgStore.GetBlobContent(imageName, manifest.Digest)
+		if err != nil {
+			imgRes := getResult(imageName, tag, errors.ErrBadBlobDigest)
+			listOfManifests[manifest.Digest] = imgRes
+
+			return imgRes
+		}
+
+		var idx ispec.Index
+		if err := json.Unmarshal(buf, &idx); err != nil {
+			imgRes := getResult(imageName, tag, errors.ErrBadBlobDigest)
+			listOfManifests[manifest.Digest] = imgRes
 
-			var idx ispec.Index
-			if err := json.Unmarshal(buf, &idx); err != nil {
-				tagName := manifest.Annotations[ispec.AnnotationRefName]
-				imgRes := getResult(imageName, tagName, errors.ErrBadBlobDigest)
-				results = append(results, imgRes)
+			return imgRes
+		}
 
-				continue
+		for _, m := range idx.Manifests {
+			imgRes := getManifestResult(ctx, m, imgStore, imageName, tag, listOfManifests)
+			listOfManifests[manifest.Digest] = imgRes
+
+			if imgRes.Error != "" {
+				return imgRes
 			}
+		}
 
-			listOfManifests = append(listOfManifests, idx.Manifests...)
-		} else if manifest.MediaType == ispec.MediaTypeImageManifest {
-			listOfManifests = append(listOfManifests, manifest)
+		okResult := getResult(imageName, tag, nil)
+
+		listOfManifests[manifest.Digest] = okResult
+
+		// check subject if exists
+		if idx.Subject != nil {
+			imgRes := getManifestResult(ctx, *idx.Subject, imgStore, imageName, tag, listOfManifests)
+			listOfManifests[manifest.Digest] = imgRes
+
+			return imgRes
 		}
-	}
 
-	for _, m := range listOfManifests {
-		tag := m.Annotations[ispec.AnnotationRefName]
-		imageResult := CheckIntegrity(ctx, imageName, tag, oci, m, dir)
-		results = append(results, imageResult)
-	}
+		return okResult
+	case ispec.MediaTypeImageManifest:
+		imgRes := CheckIntegrity(ctx, imageName, tag, manifest, imgStore)
+		listOfManifests[manifest.Digest] = imgRes
 
-	return results, nil
+		// if integrity ok then check subject if exists
+		if imgRes.Error == "" {
```

Comment:
This code should be cleaner if we don't return results, just populate the map

Reply:
Fixed.